File: python solutions/problem_15.py
#program finding the number of paths in an N by M grid
#the only moves permitted are down and right since only the shortest paths count
#a path will start at the top left corner with the destination being the bottom right corner

def grid_paths(N, M): 
    return factorial(N+M)//factorial(M)//factorial(N) #using derived formula for combination of n taken by k


def factorial(number): #function returning the factorial of a number
    if number == 2:
        return number
    else:
        return number*factorial(number-1)


if __name__=="__main__":
    print(grid_paths(20, 20))

File: haskell solutions/problem_15.hs
--Program counting all the shortest paths from a corner of a grid up to the diagonally opposite one
--A grid will be N x M with the shortest path of sum N + M 

count_shortest::Int -> Int -> Integer 
count_shortest n m = factorial (toInteger(n + m)) `div` (factorial (toInteger m)) `div` (factorial (toInteger n))


factorial::Integer -> Integer --function finding the factorial of a number 
factorial number = foldr (\x y -> x*y) 1 [1..number]


main::IO()
main = do 
    let result = map (\x -> factorial x) [1..20]
    print(result)
